test/e2e/upgrades/apps/replicasets.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/util/version"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+
+	. "github.com/onsi/ginkgo"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	interval = 2 * time.Second
+	timeout  = 1 * time.Minute
+	rsName   = "replicaset-hash-test"
+)
+
+// TODO: Test that the replicaset stays available during master (and maybe
+// node and cluster upgrades).
+
+// ReplicaSetUpgradeTest tests that a replicaset survives upgrade.
+type ReplicaSetUpgradeTest struct {
+	oldRSStatus extensions.ReplicaSetStatus
+	newRSStatus extensions.ReplicaSetStatus
+}
+
+func (ReplicaSetUpgradeTest) Name() string { return "[sig-apps] replicaset-upgrade" }
+
+func (ReplicaSetUpgradeTest) Skip(upgCtx upgrades.UpgradeContext) bool {
+	// As of 1.8, the client code we call into no longer supports talking to a server <1.7. (see #47685)
+	minVersion := version.MustParseSemantic("v1.7.0")
+
+	for _, vCtx := range upgCtx.Versions {
+		if vCtx.Version.LessThan(minVersion) {
+			return true
+		}
+	}
+	return false
+}
+
+var _ upgrades.Skippable = ReplicaSetUpgradeTest{}
+
+// Setup creates a replicaset and makes sure it has a new and an old replicaset running.
```

Comment:
I think the concept of "new" and "old" ReplicaSet came from the test for Deployment, which manages multiple ReplicaSets representing different versions of the Pod template. You shouldn't need multiple ReplicaSets here.

Reply:
Agreed. I left a comment about possibility of not having "old" RS. It suffices to run the upgrade test against a RS.